### PR TITLE
 Improve subscription performance by 10-20 times

### PR DESCRIPTION
**Pull request by MrFlashAccount**
## 🎯 Changes

This PR introduces a centralized `GCManager` that consolidates individual timeouts across queries and mutations into a single dynamically-scheduled timeout, resulting in **10-20x performance improvements** for garbage collection operations. This addresses scalability issues in applications with thousands of active queries and mutations.

**Demo**: https://react-19-query-demo-git-tanstac-d7ec69-mrflashaccounts-projects.vercel.app/

---

<details><summary>Performance difference</summary>
Before:
CleanShot 2025-10-27 at 12 53 34@2x

After: 
CleanShot 2025-10-27 at 12 52 46@2x
</details> 

---

## Problem

Previously, each query and mutation scheduled its own individual timeout for garbage collection. In applications with hundreds or thousands of queries:

- **Performance degradation**: Each query/mutation created its own `setTimeout`, leading to hundreds or thousands of timers on the event loop
- **Event loop pressure**: Excessive timers can cause performance issues and timing inaccuracies
- **Scalability limits**: As documented in the `TimeoutManager` docs, thousands of timeouts can hit platform limitations

## Solution

The new `GCManager` implements a **centralized timeout scheduling approach**:

1. **Tracking instead of scheduling**: Instead of each query/mutation scheduling its own timeout, items are marked with a timestamp (`gcMarkedAt`) when they become eligible for GC
2. **Single timeout for nearest GC**: The `GCManager` schedules a single timeout for when the nearest item becomes eligible for removal
3. **Dynamic rescheduling**: After scanning, if eligible items remain, it recalculates and reschedules the timeout for the next nearest item
4. **Automatic lifecycle**: Timeout scheduling starts when items become eligible and stops when no items remain

## Key Design Decisions

1. **Dynamic timeout scheduling**: Uses `queueMicrotask` to batch calculations and schedules a timeout for when the nearest item becomes eligible, ensuring we only scan when necessary
2. **Targeted collection**: Instead of polling with a fixed interval, the timeout precisely targets when the first item needs collection
3. **Rescheduling after scan**: After scanning, if eligible items remain, recalculates and reschedules the timeout for the next nearest item
4. **Server-side optimization**: GC is disabled on the server since there's no need for memory cleanup in server environments
5. **Backward compatible**: All existing tests pass without modification, ensuring no breaking changes

## Performance Impact

Benchmarking shows **10-20x performance improvements** for GC operations, particularly noticeable when:

- Applications have hundreds or thousands of queries/mutations
- Many queries become inactive simultaneously
- GC operations occur frequently

The centralized approach reduces:

- Timer overhead (from N timers to 1 timeout that targets the nearest item)
- Event loop pressure (only one timeout active at a time)
- Memory footprint from timeout closures (single closure per timeout instead of N)
- Unnecessary scans (only scans when items are actually ready, not on fixed intervals)

## Migration Notes

This change is **fully backward compatible**. No API changes are required for users. The improvement is transparent and automatic.

The only behavioral difference is internal: GC operations are now batched and more efficient, which may result in slightly different timing in edge cases, but the overall behavior (items being collected after their `gcTime` expires) remains identical.

---

## ✅ Checklist

- [x] I have followed the steps in the [Contributing guide](https://github.com/TanStack/query/blob/main/CONTRIBUTING.md).
- [x] I have tested this code locally with `pnpm run test:pr`.

## 🚀 Release Impact

- [x] This change affects published code, and I have generated a [changeset](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md).
- [ ] This change is docs/CI/dev-only (no release).

## Release Notes

* **New Features**
  * Implemented centralized garbage collection system that consolidates individual timeout intervals into a single scanning mechanism for improved efficiency.

* **Tests**
  * Added comprehensive test suite for garbage collection manager covering initialization, item tracking, edge cases, and error handling.
  * Updated existing tests to align with new garbage collection behavior.

### CODE REVIEW

File: packages/query-core/src/mutation.ts [hunk 1]
```diff
@@ -152,14 +158,21 @@ export class Mutation<
...
-        this.scheduleGc()
+        this.markForGc()
       } else {
         this.#mutationCache.remove(this)
+        return true
       }
     }
+
+    return false
   }
```
[the remark this answers is not in the document]
Comment (MrFlashAccount) [reply packages/query-core/src/mutation.ts:176]: Not the case as if we collect an item, GCManager removes the instance from the `#eligibleItems`

**Comment by TkDodo** [comment]
I think this is exactly why @justjake contributed the `TimeoutManager` - so that you can implement your own batching of timeouts instead of getting a new `setTimeout` for every query. 

So while these improvements are obviously good when it comes to scaling applications that use a lot of queries that gc at the same time, I don’t think I want to force certain heuristics on users. Eventually, it would be expected to have everything customizable - should we use `queueMicrotaks` or something else etc.

Have you tried passing your own `TimeoutProvider` that has the implementation you now have in this `GCManager` ?

**Comment by justjake** [comment]
Use of a timer wheel data structure can help avoid the need for O(all gc-able) scans over the set of Removable. 

Stale/refetch is another large source of timers; advantage of TimeoutManager approach is you can mitigate any/all future timer use by this package in one spot.

**Comment by justjake** [comment]
If you want to support a zillion queries efficiently you may also want to mitigate the O(n) scanning / key hashing in QueryCache, since it can be implemented as a QueryCache subclass it wasn’t merged https://github.com/TanStack/query/pull/9589

**Comment by TkDodo** [comment]
@justjake I would love to see a blogpost or even a page in the docs that outlines all the cool things you did in user-land to make TanStack Query more efficient for those cases with tons of queries ❤️

**Comment by MrFlashAccount** [comment]
> Use of a timer wheel data structure can help avoid the need for O(all GC-able) scans over the set of Removable.

@justjake  Yeah, I thought of something like this one. On the other hand, removing 2000+ different queries was almost instant. The same is true for scheduling a tick. Do you have more unique queries in your apps? 

> I think this is exactly why @justjake contributed the TimeoutManager - so that you can implement your own batching of timeouts instead of getting a new setTimeout for every query.

I think we can achieve the same with the custom timeoutManager provider, but I think this could be a breaking change (as we change the default timeoutManager) or users might override the defaults and break the perf gains. Even though we have a custom GC manager, it uses the same timeout manager under the hood, so we're compatible with custom providers. 

> I want to force certain heuristics on users.

@TkDodo wdym by this? The iml is completely opaque, and users do not need to take care of it. 

> Eventually, it would be expected to have everything customizable - should we use queueMicrotaks or something else, etc.?

Understandable. Shall we extend the provider API then?

**Comment by justjake** [comment]
@MrFlashAccount my benchmark target is 50k queries

**Comment by MrFlashAccount** [comment]
@justjake Understandable. I will improve my solution soon.

**Comment by MrFlashAccount** [comment]
@justjake, what was the purpose of the timeoutManager, it's understandable we can implement timeout batching calls in userland with this api, but what if we want to enhance it by default for all users? In such a scenario, a custom timeout provider will override the default one.

**Comment by TkDodo** [comment]
> but I think this could be a breaking change (as we change the default timeoutManager) or users might override the defaults and break the perf gains.

I didn’t mean that react-query uses a custom timeoutManager, I simply meant that you, in your code, would set a custom timeoutManager to get the behaviour you want in your code-base.

If everything is already achievable in user-land with the TimeoutManager API that we already have, I don’t want to merge this as it significantly increases bundle-size (~ 4%) for most applications that will not need this extra perf gain. Keep in mind - 99% of code bases will not have thousands of queries.

**Comment by MrFlashAccount** [comment]
4% - 500 bytes for improvement with linear scaling which means it's beneficial for 99% of users?

Even though it can be implemented it user land, why you don't want to provide this improvement for everyone?

**Comment by TkDodo** [comment]
It was 4% uncompressed, which was a couple of kb. It’s not beneficial to users if they don’t have the problem to begin with.

> Even though it can be implemented it user land, why you don't want to provide this improvement for everyone?

This is a really philosophical question that isn’t easily answerable. I wrote a blog about this some time ago: https://tkdodo.eu/blog/oss-feature-decision-tree

The bottom line is, there’s always a downside. It could be internal complexity, it could be widened API surface making it harder for newcomers or it could be feature-creep that’s outside of scope of a focussed lib.

We could add a lot of things that are beneficial to some users, but would make it worse for others. Believe it or not, the “bundle size is too big” thing comes up every month, “gc takes 150ms when I have thousands of queries” hasn’t come up once in 5 years.

So, sorry, this is’t a trade-off I’m willing to make, _especially_ because we have recently added a feature (`timeoutManager`) that also increased bundle size a bit that was made specifically so that things like this (`gcManager`) can be implemented in user-land now.

**Comment by iamakulov** [comment]
As a point of data, I’ll just note that

> gc takes 150ms when I have thousands of queries

is exactly the issue I ran into this week. (Except that it was mounting a mere hundred – not thousands! – of components, each with a `useQuery` inside.)

CleanShot 2025-11-09 at 20 34 44@2x

~For whoever it helps, here’s a userland `TimeoutProvider` that I implemented to deal with this:~ Update: you probably don’t need this. [See below.](https://github.com/TanStack/query/pull/9827#issuecomment-3530614157)

<details>

<summary>Code</summary>

```ts
import { TimeoutCallback } from "@tanstack/react-query";

/**
 * This is an implementation of `setTimeout` that batches `setTimeout` calls made with the same delay within the same microtask.
 * This improves performance, as native `setTimeout`/`clearTimeout` calls are surprisingly expensive.
 *
 * @example
 * const timeout = new MicrotaskBatchedTimeout();
 * timeout.setTimeout(() => console.log("first"), 5);
 * timeout.setTimeout(() => console.log("second"), 5);
 * // Despite two `setTimeout` calls, only one native `setTimeout` call will be made.
 *
 * const timeout = new MicrotaskBatchedTimeout({ roundTimersTo: 100 });
 * timeout.setTimeout(() => console.log("first"), 95);
 * timeout.setTimeout(() => console.log("second"), 105);
 * // Despite different delays, both callbacks will usually be called together after ~50-150ms.
 */
export class MicrotaskBatchedTimeout {
    #timerId = 0;
    #timersById = new Map<number, TimeoutCallback>();
    #timersPerTriggerTime = new Map<number, Array<number>>();
    roundTimersTo: number;
    #onError: (error: unknown) => void;

    /**
     * @param roundTimersTo - Pass a number in milliseconds to round setTimeout’s firing time to this value. E.g., setting `roundToNearestMs` to 100 and calling `setTimeout(95)` and `setTimeout(105)` will *usually* trigger a single timer, as both of these calls are within 100ms of each other. (*Usually* because it’s the timer firing time that’s rounded, not the delay.) Defaults to 1ms (= no rounding).
     * @param onError - The function to call when a timer callback throws an error. Defaults to `console.error`.
     */
    constructor({
        roundTimersTo,
        onError,
    }: {
        roundTimersTo?: number;
        onError?: (error: unknown) => void;
    } = {}) {
        this.roundTimersTo = roundTimersTo ?? 1;
        this.#onError = onError ?? console.error;
    }

    /**
     * Schedules a timer, like `setTimeout`.
     * @returns A unique timer ID that’s specific to this instance of `MicrotaskBatchedTimeout`. This is not the same as the ID returned by native `setTimeout`!
     */
    setTimeout(callback: TimeoutCallback, delay: number): number {
        const timerId = this.#timerId++;
        this.#timersById.set(timerId, callback);

        const triggerTime = this.#getTriggerTime(delay);
        const existingCallbacks = this.#timersPerTriggerTime.get(triggerTime);
        if (existingCallbacks) {
            existingCallbacks.push(timerId);
        } else {
            this.#timersPerTriggerTime.set(triggerTime, [timerId]);
            setTimeout(() => {
                this.#executeCallbacksAtTime(triggerTime);
            }, delay);
        }
        return timerId;
    }

    /**
     * Cancels a timer set by `MicrotaskBatchedTimeout.setTimeout` earlier.
     * @param timerId - The timer ID returned by `MicrotaskBatchedTimeout.setTimeout`. This is not the same as the ID returned by native `setTimeout`!
     */
    clearTimeout(timerId: number | undefined): void {
        if (timerId === undefined) return;
        this.#timersById.delete(timerId);
    }

    #executeCallbacksAtTime(triggerTime: number): void {
        const timers = this.#timersPerTriggerTime.get(triggerTime);
        if (!timers) return;
        this.#timersPerTriggerTime.delete(triggerTime);

        for (const timerId of timers) {
            const callback = this.#timersById.get(timerId);
            // The callback might be missing if the timer was cancelled with `clearTimeout`.
            if (callback) {
                try {
                    callback();
                } catch (error) {
                    this.#onError(error);
                }
            }
            this.#timersById.delete(timerId);
        }
    }

    #microtaskStartTime: number | null = null;
    #getTriggerTime(delay: number): number {
        // A microtask can last multiple milliseconds. To reliably batch `setTimeout(..., 1)` calls within the same microtask,
        // we record the microtask start time the first time `setTimeout` is called – and use it as a base time for the delay
        // until the microtask is finished.
        if (!this.#microtaskStartTime) {
            this.#microtaskStartTime = Math.floor(performance.now());
            Promise.resolve().then(() => {
                this.#microtaskStartTime = null;
            });
        }
        return (
            Math.round(
                (this.#microtaskStartTime + delay) / this.roundTimersTo,
            ) * this.roundTimersTo
        );
    }
}
```

</details>

```ts
// Usage:
const batchedTimeout = new MicrotaskBatchedTimeout({ roundTimersTo: 100 });

const timeoutProvider: TimeoutProvider<number> = {
    setTimeout: (...args) => batchedTimeout.setTimeout(...args),
    clearTimeout: (...args) => batchedTimeout.clearTimeout(...args),
    // Note: setIntervals are not batched since they are much less common in TanStack Query
    setInterval: (...args) => Number(setInterval(...args)),
    clearInterval: (...args) => clearInterval(...args),
};

timeoutManager.setTimeoutProvider(timeoutProvider);
```

**Comment by MrFlashAccount** [comment]
@iamakulov where do you batch multiple schedules of setTimeout into a single one?

**Comment by iamakulov** [comment]
> @iamakulov where do you batch multiple schedules of setTimeout into a single one?

It’s a little indirect (because it needs to support both timer rounding and cancelation) but it happens thanks to `const triggerTime = this.#getTriggerTime(delay)`:
- the first time `setTimeout()` is called in a microtask, we store the current time with `performance.now()`. this is called `microtaskStartTime`
  - all subsequent calls within the same microtask get the same `microtaskStartTime`
- any time `setTimeout()` is called, we take `microtaskStartTime`, add the passed `delay`, and then batch all timers with the same `triggerTime := microtaskStartTime + delay`
  - as a result, if you call `setTimeout(..., 40)` thrice, we’ll end up computing the identical trigger time for each of them, and they’ll end up batched
- once we have the `triggerTime`, we can batch. the batching happens with `#timersPerTriggerTime`, which maps said trigger time to all timers scheduled at that time
  - this is also indirect (as you see, it’s a map of numbers; we’re storing timer ids instead of timer callbacks directly). this is to make cancelation easy (`O(1)`)

**Comment by iamakulov** [comment]
Either way, I have run into Paul Irish, and he explained to me that nothing we do here is necessary at all.

The performance issues that we all saw (& which perhaps even led to the creation of `TimeoutManager`) **are not real**:

Screenshot 2025-11-10 at 22 08 33

In other words:
- timers are heavily affected by the profiling overhead
  - especially at the bottom of deep call stacks (such as React’s `recursivelyTraversePassiveMountEffects`), where most of these issues were observed/reported
- less obviously, timers are also (less heavily, but still significantly) affected by the mere fact of the DevTools being open
  - this is pretty wild and is the first time i’m seeing anything like this
- with DevTools closed, timers are extremely fast

You can confirm it in this benchmark: https://codepen.io/iamakulov/pen/gbrPjBr

I’m now recommending against adopting my solution above unless you can reproduce whatever performance issue you see with DevTools closed, and you’re still confident it boils down to timers.

**Comment by MrFlashAccount** [comment]
Funny how devtools tricked us
